packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+        illegalControlFlow(ctx, original->loc, "`next`");
+        return original;
+    }
+
+    ast::TreePtr preTransformReturn(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`return`");
+        return original;
+    }
+
+    ast::TreePtr preTransformRescueCase(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`rescue case`");
+        return original;
+    }
+
+    ast::TreePtr preTransformRescue(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`rescue`");
+        return original;
+    }
+
+    ast::TreePtr preTransformAssign(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`=`");
+        return original;
+    }
+
+    ast::TreePtr preTransformHash(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "hash literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformArray(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "array literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformMethodDef(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "method definitions");
+        return original;
+    }
+};
+
+// Sanity checks package files, mutates arguments to export / export_methods to point to item in namespace,
+// builds up the expression injected into packages that import the package, and codegens the <PackagedMethods>  module.
+shared_ptr<PackageInfo> getPackageInfo(core::MutableContext ctx, ast::ParsedFile &package) {
+    ENFORCE(package.file.exists());
+    ENFORCE(package.file.data(ctx).sourceType == core::File::Type::Package);
+    // Assumption: Root of AST is <root> class.
+    ENFORCE(static_cast<ast::ClassDef *>(package.tree.get()) != nullptr);
+    ENFORCE(static_cast<ast::ClassDef *>(package.tree.get())->symbol == core::Symbols::root());
+    auto packageFilePath = package.file.data(ctx).path();
+    ENFORCE(FileOps::getFileName(packageFilePath) == PACKAGE_FILE_NAME);
+    PackageInfoFinder finder;
+    package.tree = ast::TreeMap::apply(ctx, finder, move(package.tree));
+    finder.finalize(ctx);
+    if (finder.info) {
+        finder.info->packagePathPrefix = packageFilePath.substr(0, packageFilePath.find_last_of('/') + 1);
+    }
+    return move(finder.info);
+}
+
+// Add:
+//    module <PackageRegistry>::Mangled_Name_Package
+//      module ImportedPackage1
+//        # imported aliases go here
+//      end
+//    end
+// ...to __package.rb files to set up the package namespace.
+ast::ParsedFile rewritePackage(core::Context ctx, ast::ParsedFile file, const PackageInfo &package,
+                               const UnorderedMap<core::NameRef, shared_ptr<PackageInfo>> &packageInfoByMangledName) {
```

Comment:
```suggestion
                               const UnorderedMap<core::NameRef, const shared_ptr<PackageInfo>> &packageInfoByMangledName) {
```

I think that it's possible to mutate an individual PackageInfo unless you have this `const` here. Given that this is shared memory accessed concurrently, I think we really want to be as pessimistic about const-correctness as we can possibly be.

Reply:
Hmm, this might need to be a `const shared_ptr<const PackageInfo>` for that. But I think I can't do this -- I believe I'd have to copy the map from the non-const templated object into the const-templated version.

Reply:
Given that, I think it might be worthwhile to encapsulate this data into a class that we control the access to, and only allow certain operations at certain times.

I'm thinking about something similar to how we have `UnfreezeSymbolTable` and friends. I think we could make a PackagerContext, stick this in there as a private field, and then either `raise` or `ENFORCE` in the operations which write to the data structure at a time when we want to prohibit writing.

We don't have any shared mutable state like this right now, and I'd like to keep it that way to be absolutely sure we can reason easily about it as this code changes over time.

Reply:
I'm not sure I understand the concern. `PackageInfo` objects don't persist beyond the `Packager::run` function (e.g., they don't get stashed on the heap). I made it a shared ptr for convenience so I can stash it into two maps so we can look them up by name and file.

I can instead switch to a bare pointer if it makes it easier to reason about?

Reply:
There's currently nothing stopping people from accidentally writing to a `PackageInfo` object concurrently, because `rewritePackage` runs concurrently. I think that's very important to change.

Reply:
Ah! OK, I follow. I'll see if I can address this concern.

Reply:
I added a new class called `PackageDB` that facilitates thread-safe read-only access to packages. I think it addresses all of your concerns.